tests/kademlia/testprovider.nim

```diff
@@ -1,57 +1,175 @@
+
+    await kad1.bootstrap(@[switch2.peerInfo])
+
+    discard await kad1.findNode(kad2.rtable.selfId)
+
+    let
+      key1 = kad1.rtable.selfId
+      key2 = kad2.rtable.selfId
+      value = @[1.byte, 2, 3, 4, 5]
+
+    kad2.dataTable.insert(key1, value, $times.now().utc)
+    kad2.dataTable.insert(key2, value, $times.now().utc)
+
+    # ensure providermanager is empty
+    check kad1.providerManager.records.len() == 0
 
-    # kad1 has kad3 in its addressbook after adding provider
-    await kad3.addProvider(key.toCid())
+    await kad2.addProvider(key1.toCid())
+    await kad2.addProvider(key2.toCid())
     await sleepAsync(10.milliseconds)
-    check switch1.peerStore[AddressBook][switch3.peerInfo.peerId] ==
-      switch3.peerInfo.addrs
 
-    # now repeat the above without calling addProvider, should fail
-    # ensure kad1 does not have kad3 in its addressbook
-    discard switch1.peerStore[AddressBook].del(switch3.peerInfo.peerId)
-    check switch1.peerStore[AddressBook][switch3.peerInfo.peerId] !=
-      switch3.peerInfo.addrs
+    check kad1.providerManager.records.len() == 2
 
+    # wait less than expiration time
+    await sleepAsync(CleanupInterval)
+
+    # provider records have not yet expired
+    check kad1.providerManager.records.len() == 2
+
+    # wait expiration time
+    await sleepAsync(ExpirationInterval + CleanupInterval)
+
+    # provider records expired and evicted
+    check kad1.providerManager.records.len() == 0
+
+  asyncTest "Provider refreshed (not expired)":
+    var (switch1, kad1) = setupKadSwitch(PermissiveValidator(), CandSelector())
+    var (switch2, kad2) = setupKadSwitch(PermissiveValidator(), CandSelector())
+    defer:
+      await allFutures(switch1.stop(), switch2.stop())
+
+    await kad1.bootstrap(@[switch2.peerInfo])
+
+    discard await kad1.findNode(kad2.rtable.selfId)
+
+    let
+      key1 = kad1.rtable.selfId
+      key2 = kad2.rtable.selfId
+      value = @[1.byte, 2, 3, 4, 5]
+
+    kad2.dataTable.insert(key1, value, $times.now().utc)
+    kad2.dataTable.insert(key2, value, $times.now().utc)
+
+    # ensure providermanager is empty
+    check kad1.providerManager.records.len() == 0
+
+    await kad2.addProvider(key1.toCid())
+    await kad2.addProvider(key2.toCid())
+    await sleepAsync(10.milliseconds)
+
+    check kad1.providerManager.records.len() == 2
+
+    # wait less than expiration time
+    await sleepAsync(CleanupInterval)
+
+    # provider records have not yet expired
+    check kad1.providerManager.records.len() == 2
+
+    # refresh providers
+    await kad2.addProvider(key1.toCid())
+    await kad2.addProvider(key2.toCid())
+
+    # wait rest of expiration time
+    await sleepAsync(ExpirationInterval - 2 * CleanupInterval)
+
+    # provider records have not expired (refreshed)
+    check kad1.providerManager.records.len() == 2
+
+    # wait expiration time
+    await sleepAsync(ExpirationInterval + 2 * CleanupInterval)
```

Comment:
it's more correct to use kad config for these intervals, kad config is source of truth. these const are unnecessary. they also make code more unreadable as one has to trace const to see how they end up in kad config.